File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/misc/motorAccelerationLimiter.java
package org.firstinspires.ftc.teamcode.misc;

import com.qualcomm.robotcore.util.ElapsedTime;

import java.util.function.DoubleConsumer;

import static java.lang.Math.abs;
import static java.lang.Math.min;
import static java.lang.Math.signum;

public class motorAccelerationLimiter {
    private final DoubleConsumer motorToControl;
    private final ElapsedTime looptime = new ElapsedTime();
    private final double maxAcceleration;
    private double currentVelocity = 0;

    public motorAccelerationLimiter(DoubleConsumer motorToControl, double maxAcceleration) {
        this.motorToControl = motorToControl;
        this.maxAcceleration = maxAcceleration;
        looptime.reset();
    }

    public void setVelocity(double requestedVelocity) {
        if (requestedVelocity == 0) {
            currentVelocity = 0;
        } else {
            currentVelocity += min(abs(requestedVelocity - currentVelocity), abs(looptime.seconds() * maxAcceleration)) * signum(requestedVelocity - currentVelocity);
        }
        looptime.reset();
        motorToControl.accept(currentVelocity);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/misc/MotorAccelerationLimiter.kt
package org.firstinspires.ftc.teamcode.misc

import com.qualcomm.robotcore.util.ElapsedTime
import java.util.function.DoubleConsumer
import java.util.function.DoubleSupplier
import kotlin.math.abs
import kotlin.math.sign

class MotorAccelerationLimiter constructor(private val motorToControl: DoubleConsumer, private val maxAcceleration: DoubleSupplier) {

    constructor(motorToControl: DoubleConsumer,maxAcceleration: Double) : this(motorToControl, DoubleSupplier {maxAcceleration})

    private val loopTime = ElapsedTime()
    private var currentVelocity = 0.0
    fun setVelocity(requestedVelocity: Double) {
        currentVelocity += abs(requestedVelocity - currentVelocity).coerceAtMost(abs(loopTime.seconds() * maxAcceleration.asDouble)) * sign(requestedVelocity - currentVelocity)
        if (requestedVelocity == 0.0) motorToControl.accept(0.0) else motorToControl.accept(currentVelocity)
        loopTime.reset()
    }

    init {
        loopTime.reset()
    }
}